notes/shelley/Interface.md

```diff
@@ -0,0 +1,525 @@
+# Diffusion / application interface
+
+A good first step in Shelley networking implementation is to carve out the
+interface between application and diffusion, so that we can do our work
+independently (besides the architectural advantages).
+
+In summary: all "conversation" and relaying is done by the diffusion layer
+only. This way application latency does not inflate network latency. Input to
+the application layer will come from the diffusion layer rather than from a
+set of network listeners. Output to the network will go via the diffusion
+layer, and may not offer any feedback on success or failure. 
+
+A rough sketch of the interface:
+
+```Haskell
+-- Direction: application => diffusion
+
+-- Synchronously ask the diffusion layer for the blocks from a tip to a set
+-- of checkpoints. Inspired by the current block headers request that's made
+-- during recovery mode.
+getBlocks :: HeaderHash -> [HeaderHash] -> IO [Block]
+-- Special case of getting one block. May be faster to implement not in terms
+-- of getBlocks?
+getBlock :: HeaderHash -> IO (Maybe Block)
+
+-- Ask for the tip-of-chain for peers.
+-- TBD whether this will be needed.
+-- Diffusion layer decides when it returns (must have a timeout) and of course
+-- which peers to ask.
+--
+-- A 'RawBlockHeader' is given, i.e. it's decoded from the wire but not
+-- validated. Application layer can validate it as needed.
+getTips :: IO [RawBlockHeader]
+
+-- Offer a block for diffusion.
+-- Is it OK to return right away? Or should the diffusion layer respond with
+-- some indication of whether the block was actually sent to some threshold of
+-- peers?
+announceBlock :: Block -> IO ()
+
+-- Similar for other data that can be broadcast
+-- (transactions, ssc, update, etc.)
+```
+
+```Haskell
+-- Direction: diffusion => application
+
+-- Give various data to the application layer asynchronously (drop it in some
+-- queue).
+block :: Block -> IO ()
```

Comment:
Do we notify of blocks like this? I thought it was only headers, and they logic layer has to call `getBlock` or `getBlocks` to get it/them.